src/core/ddsc/src/dds_rhc_default.c

```diff
@@ -708,13 +816,18 @@ static void free_instance_rhc_free_wrap (void *vnode, void *varg)
 
 static void dds_rhc_default_free (struct dds_rhc_default *rhc)
 {
+  ddsrt_mutex_lock (&rhc->lock);
   assert (rhc_check_counts_locked (rhc, true, true));
   ddsrt_hh_enum (rhc->instances, free_instance_rhc_free_wrap, rhc);
   assert (rhc->nonempty_instances == NULL);
   ddsrt_hh_free (rhc->instances);
   lwregs_fini (&rhc->registrations);
   if (rhc->qcond_eval_samplebuf != NULL)
     ddsi_sertopic_free_sample (rhc->topic, rhc->qcond_eval_samplebuf, DDS_FREE_ALL);
+#ifdef DDSI_INCLUDE_LIFESPAN
+  lifespan_fini (&rhc->lifespan);
```

Comment:
Lifespan expiry events (from the tev thread) can still fire until this point, so you can have the tev thread blocked until you release the lock, then end up in the dds_rhc_default_sample_expired_cb, at which point stuff will break.

Calling ``dds_rhc_default_sample_expired_cb(rhc,NN_MTIME_NEVER); lifespan_fini (&rhc->lifespan);`` at the beginning should ensure that all samples with lifespan attached to them are gone and that the callback event no longer executes/will execute. (I think.)